File: code_100/java/com/interview/dynamic/SymbolExpressionEvaluation.java
package com.interview.dynamic;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Date 01/06/2016
 * @author Tushar Roy
 * Let there be a binary operation for 3 symbols a, b, c and result of these binary operation given in a table.
 * Given an expression of these 3 symbols and a final result, tell if this expression can be parenthesize in certain
 * way to produce the final result.
 *
 * Time complexity is O(n^4)
 * Space complexity is O(n^3)
 */
public class SymbolExpressionEvaluation {

    public boolean evaluateExpression(char input[][], Map<Character, Integer> index, char[] expression, char result) {
        Holder[][] T = new Holder[expression.length][expression.length];
        for (int i = 0; i < T.length; i++) {
            for (int j = 0; j < T.length; j++) {
                T[i][j] = new Holder();
            }
        }

        for (int i = 0; i < expression.length; i++) {
            T[i][i].add(expression[i]);
        }

        for (int l = 2; l <= T.length; l++) {
            for (int i = 0; i <= T.length - l; i++) {
                int j = i + l - 1;
                for (int k = i; k < j; k++) {
                    for (char ch : T[i][k].values()) {
                        for (char ch1: T[k+1][j].values()) {
                            T[i][j].add(input[index.get(ch)][index.get(ch1)]);
                        }
                    }
                }
            }
        }

        for (char ch : T[0][T.length-1].values()) {
            if ( result == ch) {
                return true;
            }
        }
        return false;
    }

    public static void main(String args[]) {
        Map<Character, Integer> index = new HashMap<>();
        index.put('a', 0);
        index.put('b', 1);
        index.put('c', 2);

        char input[][] = {{'b', 'b', 'a'}, {'c', 'b', 'a'}, {'a', 'c', 'c'}};
        SymbolExpressionEvaluation sbe = new SymbolExpressionEvaluation();
        System.out.println(sbe.evaluateExpression(input, index, "bbbbac".toCharArray(), 'a'));
    }

}

class Holder {
    private Set<Character> valueHolder = new HashSet<>();
    void add(Character ch) {
        valueHolder.add(ch);
    }
    Set<Character> values() {
        return valueHolder;
    }
}

File: showmecode_100/python/dynamic/symbolexpressionevaluation.py
"""
Problem Statement
=================

Let there be a binary operation for 3 symbols a, b, c and result of these binary operation given in a table.
Given an expression of these 3 symbols and a final result, tell if this expression can be parenthesize in certain
way to produce the final result.

Complexity
----------

* Run time Complexity: O(n^3)
* SpaceL O(n^2)

Where n is the length of the expression.
"""


def evaluate_expression(expression_map, expression, result):
    expression_length = len(expression)
    T = [[set() for _ in range(expression_length)] for _ in range(len(expression))]

    for idx, expr in enumerate(expression):
        T[idx][idx].add(expr)

    # We take a sub expression of length 2 until the total expression length
    for sub_length in range(2, expression_length + 1):
        for left_index in range(0, expression_length - sub_length + 1):
            right_index = left_index + sub_length - 1
            # we split the expression at different k indices for the total sub-expression length and store the result.
            # at T[left_index][right_index]
            # Like bbc, will be treated for (b(bc) and ((bb) c) and the final result is stored in a set at T[0][2]
            for k in range(left_index, right_index):
                for expr1 in T[left_index][k]:
                    for expr2 in T[k+1][right_index]:
                        T[left_index][right_index].add(expression_map[(expr1, expr2)])

    for expr in T[0][-1]:
        if result in expr:
            return True

    return False

if __name__ == '__main__':
    expressions = ['a', 'b', 'c']
    # expression table denotes the binary operation between two expression and its result.
    expression_table = [
        ['b', 'b', 'a'],
        ['c', 'b', 'a'],
        ['a', 'a', 'c']
    ]

    # For convenience, we can modify it to be more explicit and use the expression table

    expression_map = {
        ('a', 'a'): 'b',
        ('a', 'b'): 'b',
        ('a', 'c'): 'a',
        ('b', 'a'): 'c',
        ('b', 'b'): 'b',
        ('b', 'c'): 'a',
        ('c', 'a'): 'a',
        ('c', 'b'): 'a',
        ('c', 'c'): 'c'
    }

    assert True == evaluate_expression(expression_map, 'bbbbac', 'a')
